contributors/design-proposals/admission_control_extension.md

```diff
@@ -0,0 +1,265 @@
+core project as plugins that can be added and removed to a stock Kubernetes release. That extension
+of admission control should leverage similar mechanisms to our existing controller frameworks where possible
+and otherwise be performant and reliable.
+
+
+### Requirements
+
+1.  Easy Initialization
+    Privileged components should be able to easily participate in the **initialization** of a new object.
+2.  Synchronous Validation
+    Synchronous rejection of initialized objects or mutations must be possible outside of the kube-apiserver binary
+3.  Backwards Compatible
+    Existing API clients must see no change in behavior to external admission other than increased latency
+4.  Easy Installation
+    Administrators should be able to easily write a new admission plugin and deploy it in the cluster
+5.  Performant
+    External admission must not significantly regress performance in large and dense clusters
+6.  Reliable
+    External admission should be capable of being "production-grade" for deployment in an extremely large and dense cluster
+7.  Internally Consistent
+    Developing an admission controller should reuse as much infrastructure and tools as possible from building custom controllers so as to reduce the cost of extension.
+
+
+### Specification
+
+Based on observation of the actual admission control implementations the majority of mutation
+occurs as part of creation, and a large chunk of the remaining controllers are for side-effect free
+validation of creation and updates. Therefore we propose the following changes to Kubernetes:
+
+1.  Allow some controllers to act as "initializers" - watching the API and mutating the object before it is visible to normal clients.
+    This would reuse the majority of the infrastructure in place for controllers. Because creation is
+    one-way, the object can be "revealed" to regular clients once a set list of initializers is consumed. These
+    controllers could run on the cluster as pods.
+2.  Add a generic **external admission webhook** controller that is non-mutating (thus parallelizable)
+    This generic webhook API would resemble `admission.Interface` and be given the input object (for create) and the
+    previous object (for update). After initialization or on any update, these hooks would be invoked in parallel
+    against the remote servers and any rejection would reject the mutation.
+3.  Make the registration of both initializers and admission webhooks dynamic via the API (a configmap or cluster scoped resource)
+    Administrators should be able to dynamically add or remove hooks and initializers on demand to the cluster.
+    Configuration would be similar to registering new API group versions and include config like "fail open" or
+    "fail closed".
+
+Some admission controller types would not be possible for these extensions:
+
+* Mutating admission webhooks could be a later addition
```

Comment:
Yup, there are definitely people who would prefer a webhook. And as brendan noted, web hooks become more powerful in the age of FaaS.

That being said, our standard pattern is simple mutations of strongly consistent objects, and all of our tools and mechanisms protect that.  Admission has always been an "escape hatch".  Initializers are *not* escape hatches, they are designed behavior of the system.  I think initializers and hooks can co-exist as mentioned elsewhere in the proposal - initializers as our standard decorator, admission as the "do anything" pattern.

For the scales we are considering for OpenShift, the initializer pattern has several advantages including increased audit and observability, clear security rules, and the potential to see the changes in motion ahead of time.